### PR TITLE
Inclusion of the PACMANFrame for NDLAr and NDLArTPC type in GeoID and FragmentHeader.

### PR DESCRIPTION
**Pull request by krisfur**
Included the frame for PACMAN data from ND LAr. Added NDLArTPC to GeoID and FragmentHeader.

### CODE REVIEW

**Comment by roland-sipos** [comment]
Hi @bieryAtFnal ! This is an ND specific FrontEnd type for extending the Readout with ND specific code. It shouldn't affect any appfwk related infrastructure. Unfortunately until the move-out of the FD types are not organized and done, this is our best option to continue the ND readout development and testing.

**Review by roland-sipos — APPROVED** [review]

**Comment by philiprodrigues** [comment]
Is there a canonical reference for the PACMAN format? (In docdb or EDMS or similar). It would be good to mention that in a comment, to allow maintainers to check against the reference, update the code if the reference changes, see if there's a new version etc.

Speaking of versions, does the frame format have a version number in the frame somewhere? You will definitely want this when the format changes in future

**Comment by roland-sipos** [comment]
Hi @philiprodrigues ! Good point. The canonical definition of the PACMAN frame format does not exist yet in form of an EDMS doc. Patrick and Asher are working on those. As soon as the doc will exist, we can extend the header with the information.

On the other hand, I completely agree on the version number! @krisfur can you please comment on that?

**Comment by krisfur** [comment]
@philiprodrigues @roland-sipos I have added a version, calling this version 1. Thank you for pointing this out.  As Roland said the docs are yet to be made, but Patrick and Asher are on it now.

**Comment by roland-sipos** [comment]
@krisfur I just saw that you added a word to the header, with the version being 1 bit in it. I'm sure 1 bit won't be enough. If the structure is not defined yet with proper bit fields, at least put 5 bits as "version" in the first word.

**Comment by krisfur** [comment]
@roland-sipos made that change, but I'm still entirely unsure on how this header struct with bit fields is meant to be used, since LArPix just provided us with functions to retrieve specific information, and I don't see anything related to version in those, so I would need to a) understand what this exactly is since I have no clue about the header and b) manually create a variable for the version of the frame somehow

**Comment by krisfur** [comment]
Currently working off this: https://docs.google.com/document/d/1qYOywWfDgkKImVFdrDpT3c0idergmgzM1nPSchXhk3M/edit
Asher will discuss with Giovanna on writing a preliminary interface document.

**Comment by bieryAtFnal** [comment]
The changes to FragmentHeader and GeoID look good to me.  
I see that changes are still being made (maybe just to PACMANFrame.hpp?).
Should I add my approval to this PR or wait for more changes to be added?
Thanks,
Kurt

**Comment by krisfur** [comment]
@bieryAtFnal 
Hi Kurt, We made some last minute changes to the frame to fix some pointer dereferencing issues etc., it is all done for now and  you are free to merge. 

The only things left with the frame would be some robust timestamp setting mechanism we will need to talk to LArPix about, and will be a future pull request.
Thanks,
Krzysztof

File: include/dataformats/pacman/PACMANFrame.hpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+using word_t = uint32_t; // NOLINT(build/unsigned)
+
+/**
+ * @brief PACMAN frame
+ */
+class PACMANFrame
+{
+  public:
+  #define WORD_LEN   8  // bytes
+  #define HEADER_LEN 16 // bytes
```
Comment (eflumerf) [inline include/dataformats/pacman/PACMANFrame.hpp:28]: Use static const instead of #define

File: include/dataformats/pacman/PACMANFrame.hpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  };
+
+
+  #define MSG_TYPE_OFFSET  0 // bytes
+  #define MSG_WORDS_OFFSET 6 // bytes
+  #define UNIX_TS_OFFSET   1 // bytes
+  #define WORD_TYPE_OFFSET         0 // bytes
+  #define IO_CHANNEL_OFFSET        1 // bytes
+  #define RECEIPT_TIMESTAMP_OFFSET 4 // bytes
+  #define PACKET_OFFSET            8 // bytes
```
Comment (eflumerf) [inline include/dataformats/pacman/PACMANFrame.hpp:65]: Again, prefer variables over defines. See WIB2Frame.hpp for how it handles these "magic numbers"

File: include/dataformats/pacman/PACMANFrame.hpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+  /* ~~~ Access into message header and message contents ~~~ */
+
+  uint32_t get_msg_bytes(void* msg) {
+    // get total number of bytes in message, including header
+    return HEADER_LEN + WORD_LEN * (*get_msg_words(msg));
+  }
+
+  uint8_t* get_msg_type(void* msg) {
+    // get pointer to message type
+    return (uint8_t*)(((uint8_t*)msg) + MSG_TYPE_OFFSET);
```
Comment (eflumerf) [inline include/dataformats/pacman/PACMANFrame.hpp:77]: Better to use C++-style casts
`return static_cast<uint8_t*>(msg) + MSG_TYPE_OFFSET;`

**Comment by bieryAtFnal** [comment]
Hi Krzysztof,
I asked Eric to help look at this.  And, I see that he noted some suggestions...

**Comment by eflumerf** [comment]
As a general comment, I think it might be better to rework this code a little so that it defines overlay structs which are then applied to the given pointer or returned:

```
struct PacmanMessageHeader {
  uint32_t size_bytes;
  uint8_t message_type;
  uint16_t message_words;
  uint32_t unix_timestamp;
}

PacmanMessageHeader* get_header(void* msg) {
  return static_cast<PacmanMessageHeader*>(msg);
}

uint32_t get_msg_bytes(void* msg) {
  return get_header(msg)->size_bytes;
}
...
```

**Comment by philiprodrigues** [comment]
To be a bit more clear about my "version" comment, I was asking whether the frame format itself includes a version number, so that the format can be changed in future if necessary, in a way that allows readers to correctly interpret the frame itself (by reading the version number in the frame and then using the functions appropriate for that version). From skimming the google doc you linked, it looks like the answer is no.

So for the moment, you should be consistent with the format you're trying to interpret, which means no version in this interpreting code, but in the long term, I would suggest talking to whoever is in charge of the frame format and asking them to include a version number in the frame header.

**Comment by krisfur** [comment]
@eflumerf 
I have made and pushed the suggested changes on constexpr and C++ style casts.
As for rewriting the code to make use of a header struct, we would need to discuss with our group and likely go back to LArPix, as they originally wrote and provided the Frame class and we wouldn't feel safe at this point to make major revisions without consulting with them. 
Is it possible to merge this as it stands and for us to aim for such revisions in the future?

**Review by eflumerf — APPROVED** [review]
I think that it can be merged, but I will definitely state on the record that all of these byte and bit offset fields really make me nervous, especially about maintainability if the base data format changes. Using structs would get rid of the need to do all this pointer arithmetic as well as being much easier to understand by someone who is not intimately familiar with the data format.

**Comment by krisfur** [comment]
I agree that the readability and maintainability is an issue and will push on our group and LArPix to get some improvements on that in the (hopefully near) future.

**Comment by alessandrothea** [comment]
Hi, I strongly agree with Eric to the point that in this form the PACManFrame is de facto, not a frame and for me, it's not ready to be merged in `develop`. I see that the PR has been closed now but it's mandatory to migrate to an overlay-based-decoding for the PACManframe to be included in 2.8. @krisfur do you think it's feasible? Otherwise this PR should be rolled back.

**Comment by krisfur** [comment]
@alessandrothea I will discuss this with our working group and with LArPix and try to give you a definite answer next week. 

It would be a shame to have working code rolled back due to it not being in the desired final form, but if such is the requirement and if upon discussion we find it unfeasible to do before end of next week it would have to be the case and make us ultimately fail to meet the 2.8 deadline on a technicality.

**Pull request merged**